### ISSUE TITLE
Don't load versions from the catalog into the constraint solver that are ruled out by top-level equality constraints

### ISSUE THREAD

**Issue by glasser** [meteor/meteor#3044]
We already do this effectively for local packages but not for, eg, release packages.

**Comment by glasser** [comment]
@dgreensp is this sort of optimization something you've considered/done on your branch?

**Comment by dgreensp** [comment]
I haven't done any optimizations of that kind yet, but the code is structured in anticipation of them.  (`ConstraintSolver.Input#loadFromCatalog` currently assembles the list of all packages mentioned in top-level dependencies and constraints and calls `CatalogLoader#loadAllVersionsRecursive`, but it could easily do something different/better.)

All I need is a catalog method to call besides `catalog.getSortedVersionRecords`.  Do we still have one?

**Comment by dgreensp** [comment]
Oh I forgot to say, I did a quick measurement on an app with about 10 packages and it spent 75-200 ms loading catalog versions, which I consider significant enough to think about this.

**Comment by dgreensp** [comment]
Although, I'm not sure if optimizing a few top-level equality constraints alone will make a big dent in the amount of data loaded.

I think the API I want would let me load a range of versions, or let me ask for the sorted list of version numbers (as we had in the past) and then let me specify a list of those to fetch the records for.  Either of those interfaces would be nice.  (And maybe they exist; I haven't checked the catalog code.)

**Comment by glasser** [comment]
Well, the version solver is still structured with "first try with equality constraints for everything in `.meteor/versions`", right?  So the vast majority of the time that the version solver is run, nothing has changed, the first attempt works, and every single required package has strict equality constraints.  I would think that this would vastly cut down on load time.  (And this code runs on every rebuild.)

Also, this has transitive effects --- if you're only loading version 1.2.3 of foo, you'll never have to load any versions of a package that foo used in older versions only.

Basically, it seems to me that doing something like this would mean that in the common case (where nothing has changed and we are effectively just checking the constraints), you need to load precisely the set of versions that are actually used in the app.  (Maybe weak constraints might cause some additional ones to load.)  Which seems like a big advantage.  Especially since the next step could be to cache these particular catalog records in memory between rebuilds...

**Comment by hwillson** [comment]
Closing as this issue was opened before the Version Solver was completely re-written from scratch (Meteor 1.1), and this no longer seems to be a problem. Thanks!
